### PR TITLE
feat(cl): index CLPool.Initialize to populate sqrtPriceX96/tick pre-swap

### LINKED ISSUE #648: Investigation: 20GB OOM in processTickCrossingsForStaked — root cause + remediation plan
## Context

A 20 GB OOM crash in `processTickCrossingsForStaked` was observed at high sync depth. Investigation concluded the root cause is a structural fan-out during the preload phase of swap handlers on CL pools with staked positions — it is a property of the sweep's data-access pattern, not of any specific chain (a chain-specific batch is just where the peak RSS happens to cross the limit). This issue tracks the root-cause analysis, the evidence gathered, the interim mitigation (PR 1), and the planned structural fix.

## Root-cause model (proven from Envio source)

Chain of amplification triggered by every swap on a staked CL pool during preload:

1. **Unbounded handler concurrency.** Default `batchSize: 5000` events fire concurrently in preload with zero backpressure. Reference: `generated/node_modules/envio/src/EventProcessing.res.mjs:173-237`.
2. **LoadManager groups millions of `.get()` calls** from 5000 concurrent handlers into a single grouped DB fetch.
3. **`getWhere({ _in: [...] })` decomposes SDK-side to N separate `loadByField(\"Eq\", v)` calls**, each hitting `WHERE field = $1` — NOT a single `ANY($1::text[])` query as initially assumed. Reference: `generated/node_modules/envio/src/UserContext.res.mjs:82-86`. This is why Option 1 (batched `_in`) was rejected.
4. **`.get(id)` compiles to** `SELECT * WHERE id = ANY($1::text[])`. Reference: `generated/node_modules/envio/src/PgStorage.res.mjs:249` (`makeLoadByIdsQuery`).
5. **postgres-js text-protocol serialization of huge string arrays is memory-hostile.** `arraySerializer` (`postgres/cjs/src/types.js:241-266`) builds one contiguous JS string via `xs.map(escape).join(',')`; `bytes.js` grows-and-copies the send buffer, no streaming. At 1.77M ids × ~60 bytes UTF-16, the driver transient alone is 1–2 GB.
6. **InMemTable retains all loaded rows for the batch** with no eviction — ~0.9–1.2 GB for 1.77M `CLTickStaked` rows.
7. **Per-chain fetchState buffers.** `targetBufferSize` is `ENVIO_INDEXING_MAX_BUFFER_SIZE=5000` configured; default calc is 60k for 1 chain, 5000 for ≥6. Reference: `generated/node_modules/envio/src/ChainManager.res.mjs:22-40`, `generated/node_modules/envio/src/Env.res.mjs:17,87`.

Combined: 1–2 GB driver + 1–2 GB InMemTable + 5000 handler closures + V8 overhead + cross-batch fetchState buffers ≈ 20 GB plausible peak.

### Benchmark + measurement scripts (in `scripts/`)

- `scripts/bench-tick-crossings.ts` — Layer 1 microbench (three variants: baseline / PR 1 / Option 1). **Note:** its Option-1 numbers (4.9 GB → 622 MB) are artifacts of the mock treating `getWhere` as O(1); they do not reflect real Envio behavior. Kept as a historical reference for the investigation.
- `scripts/measure-cltickstaked-max.ts` — measures the worst-case N (`CLTickStaked` rows per pool) on the deployed indexer.
- `scripts/measure-staked-nfp-per-pool.ts` — measures the distribution of currently-staked `NonFungiblePosition` entities per pool (sizes K for the planned tick-edge list).
- `scripts/measure-cl-pool-staked-vs-swaps.ts` — measures the true staked vs unstaked CL pool distribution and swap weighting.
- `scripts/measure-negative-pool-fields.ts` — audits deployed pools for negative staked counters (drift from ordering bugs / sign errors).

### Envio / postgres-js source citations

- `generated/node_modules/envio/src/UserContext.res.mjs:82-86` — `_in` decomposition into N loadByField calls
- `generated/node_modules/envio/src/PgStorage.res.mjs:249` — `makeLoadByIdsQuery` compiles `.get()` to `ANY($1::text[])`
- `generated/node_modules/envio/src/PgStorage.res.mjs:245` — `makeLoadByFieldQuery` compiles `getWhere` to `WHERE \"field\" operator $1`
- `generated/node_modules/envio/src/EventProcessing.res.mjs:173-237` — unbounded preload concurrency
- `generated/node_modules/envio/src/ChainManager.res.mjs:22-40` — per-chain `targetBufferSize` calculation
- `generated/node_modules/envio/src/Env.res.mjs:17,87` — `ENVIO_INDEXING_MAX_BUFFER_SIZE`, `ENVIO_PG_MAX_CONNECTIONS`
- postgres-js `types.js:241-266`, `bytes.js` — text-protocol array serialization + send-buffer grow-and-copy

### PR 1 — hasStakes latch + bounds clamp (interim hotfix)

Scope (in `src/Aggregators/CLStakedLiquidity.ts` + supporting handlers/tests):

- `hasStakes=false` short-circuits the entire sweep for pools that have never been staked.
- Ticks outside `[TICK_MIN=-887272n, TICK_MAX=887272n]` are rejected and logged rather than triggering unbounded iteration.

This handles the unstaked-pool slice of the workload (per `measure-cl-pool-staked-vs-swaps.ts` data) but does not remove the structural fan-out on staked pools — which is the core OOM driver.

Tracked in child issue #650. Implementation opened as PR #651.

### Option 2 — sparse tick-edge list on `LiquidityPoolAggregator` (structural fix, planned)

Tracked in child issue #649.

Replace the fan-out pattern with an in-aggregator sorted list of tick edges:

- Add `stakedTickEdges: [BigInt!]!` (sorted, monotone) to `LiquidityPoolAggregator`.
- Maintain the list on gauge deposit / withdraw (append / remove) and NFPM liq-change events. Affected modules: `src/EventHandlers/Gauges/GaugeSharedLogic.ts`, NFPM common logic.
- Rewrite `processTickCrossingsForStaked` to binary-search the in-memory edge array. No `CLTickStaked.get` calls on the swap path.

**Why this eliminates each OOM mechanism:**

| Mechanism                                                 | Eliminated by Option 2 because…                                       |
| --------------------------------------------------------- | --------------------------------------------------------------------- |
| 5000 concurrent handler fan-out                           | no new `.get()` calls issued from the swap path                       |
| LoadManager grouping millions of ids                      | no `.get(CLTickStaked)` calls → no LoadManager growth                 |
| 1–2 GB driver text[] serialization                        | no `ANY` query fires for tick crossings                               |
| 0.9–1.2 GB InMemTable `CLTickStaked` accumulation         | no `CLTickStaked` rows loaded                                         |
| `addEmptyIndex` O(N) per `getWhere`                       | N/A — path uses neither `.get()` nor `.getWhere()` for tick crossings |

**Worst-case cost** (1:1 gauge:swap ratio, E = 22k tick edges per pool): each write is an O(E) immutable array copy ≈ 50µs at E=22k (V8 memcpy of ~176KB pointers at ~10 GB/s). At 2500 writes/batch, that's ~125ms extra CPU per batch — vs a 20 GB OOM. Peak extra memory: one aggregator version in flight ≈ 1.4MB (old GC'd between writes). InMemTable stores only the latest version = one commit write per pool regardless of event count.

### Rejected / deferred

- **Option 1 (batched `_in`).** Rejected. `UserContext.res.mjs:82-86` proves `_in` decomposes to N separate `WHERE field = $1` queries, strictly worse than a single `ANY`.
- **`fullBatchSize: 1000` config stopgap.** Dropped — redundant once Option 2 ships. Only useful as a bridge if the structural fix took longer than acceptable.
- **Standalone Layer 3 MockDb harness.** Dropped as a separate workstream. The correctness argument for Option 2 is structural (chain-of-amplification broken at every link); a co-located sanity test inside the Option 2 PR covers the maintenance invariants.

## Child issues

- PR 1 (interim hotfix): #650 — opened as PR #651
- Option 2 (structural fix): #649

### LINKED ISSUE #654: Index CLPool.Initialize to populate sqrtPriceX96/tick before the first Swap
## Summary

`CLPool.Initialize(sqrtPriceX96, tick)` is not currently indexed. The schema already documents the intent — `schema.graphql:48-49` annotates `LiquidityPoolAggregator.sqrtPriceX96` and `tick` with `updated from Swap/Initialize events` — but only `Swap` actually writes them. This leaves a lifecycle dead-zone between `pool.initialize()` and the first `pool.swap()` where the aggregator's `sqrtPriceX96` is `0n` and downstream code bails out.

## Current bail-out sites

Two NFPM handler paths silently drop the stake/range calculation when `sqrtPriceX96 === 0n`:

- `src/EventHandlers/NFPM/NFPMCommonLogic.ts:152` — `updateStakedPositionLiquidity`: early-returns before computing `stakedLiquidityInRange`, `stakedReserve0/1`. As of PR #653 the `stakedTickEdges`/`stakedTickEdgeNets` arrays still get updated on this path, but the staked-in-range/reserves numbers stay at their prior (possibly 0) value.
- `src/EventHandlers/NFPM/NFPMTransferLogic.ts:213` — NFPM Transfer handler: no-ops entirely for mint/transfer attribution if the pool hasn't had a swap yet.

These fire when liquidity is minted / transferred on a pool *after* `initialize()` but *before* the first `swap()` — a plausible new-pool-seeding sequence.

## Fields that would become correct earlier in pool lifecycle

Directly populated by `Initialize`:
- `LiquidityPoolAggregator.sqrtPriceX96`
- `LiquidityPoolAggregator.tick`
- `LiquidityPoolAggregatorSnapshot.sqrtPriceX96`
- `LiquidityPoolAggregatorSnapshot.tick`

Indirectly corrected (no longer dropped by the `0n` bail on pre-first-swap NFPM events):
- `LiquidityPoolAggregator.stakedLiquidityInRange`
- `LiquidityPoolAggregator.stakedReserve0`
- `LiquidityPoolAggregator.stakedReserve1`
- `UserStatsPerPool.*` attributions via `NFPMTransferLogic` for pre-swap mints

## Proposal

Two-step, measurement-first:

1. **Measure**. One-off script: for each CL pool, find the earliest `NFPM.IncreaseLiquidity` and the earliest `CLPool.Swap`. Count pools where the IncreaseLiquidity predates the Swap, and count how many positions land in that window. If the count is trivial (say <50 positions across all chains) the bail-out is benign and we can close this issue as not-needed.
2. **Implement** (only if step 1 shows a meaningful gap). Add `CLPool.Initialize` to `config.yaml` events, add a handler that sets `sqrtPriceX96` and `tick` on the aggregator at the `Initialize` block, and drop or narrow the `sqrtPriceX96 === 0n` bail-outs.

## Out of scope

- Back-fill for historical pre-swap positions. The indexer is forward-looking; re-running from genesis is the way to reprocess if we care about historical correctness.
- Any change to swap-path or staked-edge logic (PR #653 already finalized those).

## Context

- Surfaced during review of PR [#653](https://github.com/velodrome-finance/indexer/pull/653) (CLTickStaked fan-out -> sparse edges list).
- Related: #648 (OOM), #649 (swap-path optimization), #652 (legacy CLTickStaked removal).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by FonsSicca**
Closes #654. Stacks on #653 (`feature/cl-staked-tick-edges-sparse-list`) — base this PR on #653 and rebase onto `main` once #653 merges.

## Summary

- Subscribe `CLPool` to `Initialize` in `config.yaml`
- New `processCLPoolInitialize` logic module (`src/EventHandlers/CLPool/CLPoolInitializeLogic.ts`) producing a `{sqrtPriceX96, tick, lastUpdatedTimestamp}` aggregator diff
- Register `CLPool.Initialize.handler` in `src/EventHandlers/CLPool.ts`, slotted alphabetically between `IncreaseObservationCardinalityNext` and `Mint`
- Narrow the two `sqrtPriceX96 === 0n` bail-outs (`NFPMCommonLogic.ts:152`, `NFPMTransferLogic.ts:213`) from "primary path" to "defensive fallback for pre-existing data where Initialize was never indexed", with comments pointing at #654

## Why

`CLPool.Initialize(sqrtPriceX96, tick)` was not indexed. `schema.graphql:48-49` already documented the intent (`sqrtPriceX96` / `tick` "updated from Swap/Initialize events") but only `Swap` wrote the fields. Between `pool.initialize()` and the first `pool.swap()`, the aggregator kept `sqrtPriceX96 === 0n`, causing:

- `NFPMCommonLogic.updateStakedPositionLiquidity` to skip `stakedLiquidityInRange` / `stakedReserve0/1` math for pre-swap mints (#653 kept the edge-list write — just not the reserves/in-range number)
- `NFPMTransferLogic.attributeTransferToUserStatsPerPool` to no-op entirely for mint/transfer attribution on pre-swap positions

Deviation from issue's step 1 (measurement): skipped because the measurement needs live DB access the author does not have, the fix is small and self-contained, and the schema already treats Initialize as an authoritative update source. Bail-outs stay (narrowed) so pre-existing data from older indexer runs does not explode when replayed.

## Test plan
- [x] `pnpm qa --write` clean
- [x] `pnpm tsc --noEmit` clean
- [x] `pnpm test` — 97/97 test files pass
- [x] New tests (`test/EventHandlers/CLPool/CLPoolInitializeLogic.test.ts`): unit + `processEvents` integration, negative-tick int24 preservation, unknown-pool defensive no-op
- [ ] Reviewer spot-check: on a fresh re-index, confirm the first CLPool.Initialize event updates `LiquidityPoolAggregator.sqrtPriceX96` and `tick` before the first Swap

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added indexing support for CLPool `Initialize` events to properly track when liquidity pools are initialized.

* **Documentation**
  * Enhanced clarifications regarding uninitialized pool handling to prevent incorrect token amount calculations in edge cases.

### CODE REVIEW

**Review by WooSungD — APPROVED** [review]

**Comment by FonsSicca** [comment]
@coderabbitai review

File: test/EventHandlers/CLPool/CLPoolInitializeLogic.test.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        },
+      });
+
+      // Should not throw or create a phantom aggregator for the unknown pool.
+      const result = await mockDb.processEvents([mockEvent]);
+      expect(
+        result.entities.LiquidityPoolAggregator.get(
+          `${chainId}-${unknownPool}`,
+        ),
+      ).toBeUndefined();
```
[the remark this answers is not in the document]
Comment (FonsSicca) [reply test/EventHandlers/CLPool/CLPoolInitializeLogic.test.ts:134]: > Use the `PoolId()` helper instead of hand-building the key string.

Addressed in b9c152b: imported `PoolId` from `src/Constants.ts` and replaced the hand-built key with `PoolId(chainId, unknownPool)` so the test stays aligned with the canonical format.
